src/Propulsion.EventStore/EventStoreSource.fs

```diff
@@ -0,0 +1,98 @@
+module Mapping =
+    open EventStore.ClientAPI
+
+    type RecordedEvent with
+        member __.Timestamp = DateTimeOffset.FromUnixTimeMilliseconds(__.CreatedEpoch)
+
+    let (|PropulsionEvent|) (x : RecordedEvent) =
+        { new Propulsion.Streams.IEvent<_> with
+            member __.EventType = x.EventType
+            member __.Data = if x.Data <> null && x.Data.Length = 0 then null else x.Data
+            member __.Meta = if x.Metadata <> null && x.Metadata.Length = 0 then null else x.Metadata
+            member __.Timestamp = x.Timestamp }
+    let (|PropulsionStreamEvent|) (x: RecordedEvent) : Propulsion.Streams.StreamEvent<_> =
+        { stream = x.EventStreamId; index = x.EventNumber; event = (|PropulsionEvent|) x }
+
+type EventStoreSource =
+    static member Run
+        (   log : Serilog.ILogger, sink : Propulsion.ProjectorPipeline<_>, checkpoints : Checkpoint.CheckpointSeries,
+            connect, spec, categorize, tryMapEvent,
+            maxReadAhead, statsInterval) = async {
+        let conn = connect ()
+        let! maxInParallel = Async.StartChild <| Reader.establishMax conn
+        let! initialCheckpointState = checkpoints.Read
+        let! maxPos = maxInParallel
+        let! startPos = async {
+            let mkPos x = EventStore.ClientAPI.Position(x, 0L)
+            let requestedStartPos =
+                match spec.start with
+                | Absolute p -> mkPos p
+                | Chunk c -> Reader.posFromChunk c
+                | Percentage pct -> Reader.posFromPercentage (pct, maxPos)
+                | TailOrCheckpoint -> maxPos
+                | StartOrCheckpoint -> EventStore.ClientAPI.Position.Start
+            let! startMode, startPos, checkpointFreq = async {
+                match initialCheckpointState, requestedStartPos with
+                | Checkpoint.Folds.NotStarted, r ->
+                    if spec.forceRestart then invalidOp "Cannot specify --forceRestart when no progress yet committed"
+                    do! checkpoints.Start(spec.checkpointInterval, r.CommitPosition)
+                    return Starting, r, spec.checkpointInterval
+                | Checkpoint.Folds.Running s, _ when not spec.forceRestart ->
+                    return Resuming, mkPos s.state.pos, TimeSpan.FromSeconds(float s.config.checkpointFreqS)
+                | Checkpoint.Folds.Running _, r ->
+                    do! checkpoints.Override(spec.checkpointInterval, r.CommitPosition)
+                    return Overridding, r, spec.checkpointInterval }
+            log.Information("Sync {mode} {groupName} @ {pos} (chunk {chunk}, {pct:p1}) checkpointing every {checkpointFreq:n1}m",
+                startMode, spec.groupName, startPos.CommitPosition, Reader.chunk startPos, float startPos.CommitPosition/float maxPos.CommitPosition,
+                checkpointFreq.TotalMinutes)
+            return startPos }
+        let cosmosIngester = sink.StartIngester(log.ForContext("Tranche","Ingester"), 0)
```

Comment:
is this a typo, or does the EventStoreSource really need Cosmos?

Reply:
Good catch - before it was generalized to sync to/from ES/Cosmos, the original impl was indeed ES->Cosmos. `ingester` would be a better name, as really all a source does is feed into _an_ ingester. (TMI: an ingester feeds to a scheduler - which then dispatches to a handler (which might be a sink if the task is to maintain a live sync between two event storage systems))